### PR TITLE
Added option to AreIn to perform a strict test and return false if the collection is null or empty

### PR DESCRIPTION
**Pull request by shlokam**
#280

### CODE REVIEW

**Comment by shlokam** [comment]
@t92549 Should I write test cases for the above scenario ?

File: core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java [hunk 1]
```diff
@@ -80,6 +82,10 @@ public void setAllowedValues(final Object[] allowedValuesArray) {
 
     @Override
     public boolean test(final Collection<?> input) {
+        if (isNullOrEmptyAllowed) {
+            return (!(null == allowedValues || allowedValues.isEmpty()));
```
Comment (t92549) [inline core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java:86]: Instead of a separate return statement, this logic could be added into the one below. An example for the null bit:
`return (allowedValues == null && isNullOrEmptyAllowed) || ...`
These may need to be put onto newlines as it will get quite long

File: core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java [hunk 2]
```diff
@@ -40,13 +40,15 @@
 @Summary("Checks if a provided collection contains all the provided input values")
 public class AreIn extends KoryphePredicate<Collection<?>> {
     private Collection<?> allowedValues;
+    private boolean isNullOrEmptyAllowed;
```
Comment (t92549) [inline core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java:43]: This should default to previous behaviour, which is true

**Review by t92549 — CHANGES_REQUESTED** [review]

**Comment by t92549** [comment]
> @t92549 Should I write test cases for the above scenario ?

Yes please, a test like `AreInTest::shouldAcceptWhenValuesIsEmpty` which uses the new flag to assert it should reject the list

File: core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java (hunk 1 above)
Comment (shlokam) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java]: I added a separate return Statement because writing 

(return (allowedValues == null && isNullOrEmptyAllowed) || ...)

was allowing empty input to return true even if flag was false.

File: core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java (hunk 2 above)
Comment (shlokam) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java]: Added

**Comment by shlokam** [comment]
> > @t92549 Should I write test cases for the above scenario ?
> 
> Yes please, a test like `AreInTest::shouldAcceptWhenValuesIsEmpty` which uses the new flag to assert it should reject the list

Added

File: core/src/test/java/uk/gov/gchq/koryphe/impl/predicate/AreInTest.java [hunk 3]
```diff
@@ -166,4 +166,26 @@ protected Iterable<AreIn> getDifferentInstancesOrNull() {
                 new AreIn(VALUE2)
         );
     }
+
+    @Test
+    public void shouldRejectWhenValuesIsEmptyAndEmptyNotAllowed() {
```
Comment (t92549) [inline core/src/test/java/uk/gov/gchq/koryphe/impl/predicate/AreInTest.java:171]: Maybe more tests with null being used as well as empty list would be good

File: core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java [hunk 4]
```diff
@@ -80,7 +82,11 @@ public void setAllowedValues(final Object[] allowedValuesArray) {
 
     @Override
     public boolean test(final Collection<?> input) {
-        return null == allowedValues || allowedValues.isEmpty() || (null != input && allowedValues.containsAll(input));
+        if (null == allowedValues || allowedValues.isEmpty()) {
+            return (isNullOrEmptyAllowed);
```
Comment (t92549) [inline core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java:86]: I don't think these brackets are needed

**Review by t92549 — CHANGES_REQUESTED** [review]
Nice changes! Just a couple comments

File: core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java (hunk 4 above)
Comment (GCHQDeveloper314) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java:86]: I think this is specifically the `return (isNullOrEmptyAllowed);` should be `return isNullOrEmptyAllowed;` and not the `if` where they brackets are required.

File: core/src/test/java/uk/gov/gchq/koryphe/impl/predicate/AreInTest.java (hunk 3 above)
Comment (shlokam) [reply core/src/test/java/uk/gov/gchq/koryphe/impl/predicate/AreInTest.java:171]: Ok. Will add that

File: core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java (hunk 4 above)
Comment (shlokam) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/predicate/AreIn.java:86]: Ok. will change that.

**Comment by GCHQDeveloper314** [comment]
The varargs constructor doesn't take the `isNullOrEmptyAllowed` option. Therefore trying to pass in multiple collections while not allowing them to be null/empty isn't possible.
This could be fixed by adding check for a `Boolean` as the last item in the varargs list.

**Comment by shlokam** [comment]
> The varargs constructor doesn't take the `isNullOrEmptyAllowed` option. Therefore trying to pass in multiple collections while not allowing them to be null/empty isn't possible. This could be fixed by adding check for a `Boolean` as the last item in the varargs list.

Ok. Will add test cases also for the same.

**Comment by shlokam** [comment]
Can you tell why this failed?

**Comment by t92549** [comment]
> Can you tell why this failed?

It failed due to code coverage reduction. I have submitted a cleanup commit, part of which removes the vararg feature requested as I think it may casue bugs with collections of Booleans.

**Review by GCHQDeveloper314 — APPROVED** [review]

**Review by t92549 — APPROVED** [review]

**Comment by shlokam** [comment]
> > Can you tell why this failed?
> 
> It failed due to code coverage reduction. I have submitted a cleanup commit, part of which removes the vararg feature requested as I think it may cause bugs with collections of Booleans.

Ok. Thank you.
Can you add the hacktoberfest-accepted to the PR?
